File: sdk/node/src/retry-logic.ts
/**
 * Retry Logic with Exponential Backoff for WorkerSQL
 * Handles automatic retries for transient errors
 */

import { ValidationError } from '../../schema/validator.js';

export interface RetryOptions {
  maxAttempts?: number;
  initialDelay?: number;
  maxDelay?: number;
  backoffMultiplier?: number;
  retryableErrors?: string[];
}

const DEFAULT_RETRYABLE_ERRORS = [
  'CONNECTION_ERROR',
  'TIMEOUT_ERROR',
  'RESOURCE_LIMIT',
  'ECONNREFUSED',
  'ECONNRESET',
  'ETIMEDOUT',
  'ENETUNREACH',
];

export class RetryStrategy {
  private options: Required<RetryOptions>;

  constructor(options: RetryOptions = {}) {
    this.options = {
      maxAttempts: options.maxAttempts ?? 3,
      initialDelay: options.initialDelay ?? 1000,
      maxDelay: options.maxDelay ?? 30000,
      backoffMultiplier: options.backoffMultiplier ?? 2,
      retryableErrors: options.retryableErrors ?? DEFAULT_RETRYABLE_ERRORS,
    };
  }

  /**
   * Check if an error is retryable
   */
  isRetryable(error: any): boolean {
    if (error instanceof ValidationError) {
      return this.options.retryableErrors.includes(error.code);
    }

    if (error?.code && typeof error.code === 'string') {
      return this.options.retryableErrors.includes(error.code);
    }

    if (error?.message && typeof error.message === 'string') {
      return this.options.retryableErrors.some(code => 
        error.message.includes(code)
      );
    }

    return false;
  }

  /**
   * Calculate delay for a given attempt
   */
  calculateDelay(attempt: number): number {
    const delay = this.options.initialDelay * Math.pow(this.options.backoffMultiplier, attempt);
    return Math.min(delay, this.options.maxDelay);
  }

  /**
   * Add jitter to prevent thundering herd
   */
  addJitter(delay: number): number {
    const jitter = Math.random() * 0.3 * delay; // Up to 30% jitter
    return delay + jitter;
  }

  /**
   * Execute a function with retry logic
   */
  async execute<T>(
    fn: () => Promise<T>,
    context?: string
  ): Promise<T> {
    let lastError: any;

    for (let attempt = 0; attempt < this.options.maxAttempts; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error;

        // Check if we should retry
        if (!this.isRetryable(error)) {
          throw error;
        }

        // Check if we've exhausted retries
        if (attempt === this.options.maxAttempts - 1) {
          throw new ValidationError(
            'CONNECTION_ERROR',
            `Failed after ${this.options.maxAttempts} attempts${context ? ` (${context})` : ''}`,
            { originalError: error, attempts: attempt + 1 }
          );
        }

        // Calculate and apply delay
        const delay = this.calculateDelay(attempt);
        const delayWithJitter = this.addJitter(delay);

        console.debug(
          `[WorkerSQL Retry] Attempt ${attempt + 1}/${this.options.maxAttempts} failed${context ? ` (${context})` : ''}. ` +
          `Retrying in ${Math.round(delayWithJitter)}ms...`
        );

        await new Promise(resolve => setTimeout(resolve, delayWithJitter));
      }
    }

    throw lastError;
  }

  /**
   * Execute with timeout
   */
  async executeWithTimeout<T>(
    fn: () => Promise<T>,
    timeout: number,
    context?: string
  ): Promise<T> {
    return Promise.race([
      this.execute(fn, context),
      new Promise<T>((_, reject) =>
        setTimeout(
          () => reject(new ValidationError('TIMEOUT_ERROR', `Operation timed out after ${timeout}ms${context ? ` (${context})` : ''}`)),
          timeout
        )
      ),
    ]);
  }
}

File: sdk/python/workersql_client/retry_logic.py
"""
Retry Logic with Exponential Backoff for WorkerSQL
Handles automatic retries for transient errors
"""

import random
import time
from typing import Any, Callable, List, Optional, TypeVar

from .workersql_client import ValidationError

T = TypeVar("T")

DEFAULT_RETRYABLE_ERRORS = [
    "CONNECTION_ERROR",
    "TIMEOUT_ERROR",
    "RESOURCE_LIMIT",
]


class RetryStrategy:
    """Retry strategy with exponential backoff"""

    def __init__(
        self,
        max_attempts: int = 3,
        initial_delay: float = 1.0,
        max_delay: float = 30.0,
        backoff_multiplier: float = 2.0,
        retryable_errors: Optional[List[str]] = None,
    ):
        """
        Initialize retry strategy

        Args:
            max_attempts: Maximum number of retry attempts
            initial_delay: Initial delay in seconds
            max_delay: Maximum delay in seconds
            backoff_multiplier: Multiplier for exponential backoff
            retryable_errors: List of error codes that should trigger a retry
        """
        self.max_attempts = max_attempts
        self.initial_delay = initial_delay
        self.max_delay = max_delay
        self.backoff_multiplier = backoff_multiplier
        self.retryable_errors = retryable_errors or DEFAULT_RETRYABLE_ERRORS

    def is_retryable(self, error: Exception) -> bool:
        """
        Check if an error is retryable

        Args:
            error: The exception to check

        Returns:
            bool: True if the error should trigger a retry
        """
        if isinstance(error, ValidationError):
            return error.code in self.retryable_errors

        # Check for common network/connection errors
        error_msg = str(error).lower()
        network_errors = [
            "connection",
            "timeout",
            "refused",
            "reset",
            "unreachable",
        ]
        return any(err in error_msg for err in network_errors)

    def calculate_delay(self, attempt: int) -> float:
        """
        Calculate delay for a given attempt

        Args:
            attempt: Current attempt number (0-indexed)

        Returns:
            float: Delay in seconds
        """
        delay = self.initial_delay * (self.backoff_multiplier**attempt)
        return min(delay, self.max_delay)

    def add_jitter(self, delay: float) -> float:
        """
        Add jitter to prevent thundering herd

        Args:
            delay: Base delay in seconds

        Returns:
            float: Delay with jitter added
        """
        jitter = random.random() * 0.3 * delay  # Up to 30% jitter
        return delay + jitter

    def execute(self, fn: Callable[[], T], context: Optional[str] = None) -> T:
        """
        Execute a function with retry logic

        Args:
            fn: Function to execute
            context: Optional context string for logging

        Returns:
            Result of the function

        Raises:
            ValidationError: If all retry attempts fail
        """
        last_error: Optional[Exception] = None

        for attempt in range(self.max_attempts):
            try:
                return fn()
            except Exception as error:
                last_error = error

                # Check if we should retry
                if not self.is_retryable(error):
                    raise

                # Check if we've exhausted retries
                if attempt == self.max_attempts - 1:
                    context_str = f" ({context})" if context else ""
                    raise ValidationError(
                        "CONNECTION_ERROR",
                        f"Failed after {self.max_attempts} attempts{context_str}",
                        {"original_error": str(error), "attempts": attempt + 1},
                    )

                # Calculate and apply delay
                delay = self.calculate_delay(attempt)
                delay_with_jitter = self.add_jitter(delay)

                context_str = f" ({context})" if context else ""
                print(
                    f"[WorkerSQL Retry] Attempt {attempt + 1}/{self.max_attempts} failed{context_str}. "
                    f"Retrying in {delay_with_jitter:.2f}s..."
                )

                time.sleep(delay_with_jitter)

        # This should never be reached, but just in case
        if last_error:
            raise last_error
        raise ValidationError("INTERNAL_ERROR", "Unexpected retry failure")
